### PR TITLE
feat(proxy-rewrite): create use_real_request_uri_unsafe option

### LINKED ISSUE #7386: feat: Expose an option for (real_)request_uri to be used in proxy-rewrite plugin
### Description

Hello!

During our tests we have noticed that APISIX always uses the normalized/decoded URI coming from nginx (as per its design, which isn't an issue), and this creates an edge case where a backend that _requires_ URL Encoded URIs to not work properly (i.e. GitLab API calls that reference paths inside a project use `%2F` instead of `/`).

Looking at the codebase, we've also seen the native `request_uri` variable being moved to the `real_request_uri` variable. And because of how `proxy-rewrite`s rewrite schema work; utilizing `real_request_uri` to bypass the encoding is not possible.

We know that this is a really unsafe behavior to utilize `real_request_uri` but the cost of the edge cases this causes is critical. We've copied over `proxy-rewrite` to [our downstream custom plugins repo as `proxy-rewrite-unsafe`](https://github.com/Trendyol/apisix-plugins/blob/main/plugins/proxy-rewrite-unsafe/proxy-rewrite-unsafe.lua) with our addition of the `real_request_uri` override:

```diff
--- proxy-rewrite.lua	2022-07-05 09:49:53.000000000 +0300
+++ proxy-rewrite-unsafe.lua	2022-07-05 10:02:39.000000000 +0300
@@ -15,7 +15,7 @@
 -- limitations under the License.
 --
 local core        = require("apisix.core")
-local plugin_name = "proxy-rewrite"
+local plugin_name = "proxy-rewrite-unsafe"
 local pairs       = pairs
 local ipairs      = ipairs
 local ngx         = ngx
@@ -40,6 +40,11 @@
 local schema = {
     type = "object",
     properties = {
+        use_real_request_uri_unsafe = {
+            description = "use real_request_uri (request_uri in nginx), THIS IS UNSAFE.)",
+            type        = "boolean",
+            default     = false,
+        },
         uri = {
             description = "new uri for upstream",
             type        = "string",
@@ -161,7 +166,9 @@
     end
 
     local upstream_uri = ctx.var.uri
-    if conf.uri ~= nil then
+    if conf.use_real_request_uri_unsafe then
+        upstream_uri = core.utils.resolve_var("$real_request_uri", ctx.var)
+    elseif conf.uri ~= nil then
         upstream_uri = core.utils.resolve_var(conf.uri, ctx.var)
     elseif conf.regex_uri ~= nil then
         local uri, _, err = re_sub(ctx.var.uri, conf.regex_uri[1],
@@ -178,18 +185,22 @@
     end
 
     local index = str_find(upstream_uri, "?")
-    if index then
-        upstream_uri = core.utils.uri_safe_encode(sub_str(upstream_uri, 1, index-1)) ..
-                       sub_str(upstream_uri, index)
-    else
-        upstream_uri = core.utils.uri_safe_encode(upstream_uri)
+    if not conf.use_real_request_uri_unsafe then
+        if index then
+            upstream_uri = core.utils.uri_safe_encode(sub_str(upstream_uri, 1, index-1)) ..
+                           sub_str(upstream_uri, index)
+        else
+            upstream_uri = core.utils.uri_safe_encode(upstream_uri)
+        end
     end
 
     if ctx.var.is_args == "?" then
-        if index then
-            ctx.var.upstream_uri = upstream_uri .. "&" .. (ctx.var.args or "")
-        else
-            ctx.var.upstream_uri = upstream_uri .. "?" .. (ctx.var.args or "")
+        if not conf.use_real_request_uri_unsafe then
+            if index then
+                ctx.var.upstream_uri = upstream_uri .. "&" .. (ctx.var.args or "")
+            else
+                ctx.var.upstream_uri = upstream_uri .. "?" .. (ctx.var.args or "")
+            end
         end
     else
         ctx.var.upstream_uri = upstream_uri

```

Is it possible to expose an option on upstream like in the diff above?

### PR DESCRIPTION
**Pull request by ilteriseroglu-ty**
### Description

This PR fixes #7386.

As per the linked issue:

> During our tests we have noticed that APISIX always uses the normalized/decoded URI coming from nginx (as per its design, which isn't an issue), and this creates an edge case where a backend that _requires_ URL Encoded URIs to not work properly (i.e. GitLab API calls that reference paths inside a project use `%2F` instead of `/`).

This PR implements an option named `use_real_request_uri_unsafe` to utilize `real_request_uri` (which is originally `request_uri` in nginx; its the request uri that's _not_ normalized) and bypass all encoding calls inside.

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Review by tokers — COMMENTED** [review]
LGTM

File: t/lib/server.lua [hunk 1]
```diff
@@ -377,6 +377,13 @@ for i = 1, 100 do
     _M["print_uri_" .. i] = print_uri
 end
 
+local function print_uri_detailed()
+    ngx.say("ngx.var.uri: ", ngx.var.uri)
+    ngx.say("ngx.var.request_uri: ", ngx.var.request_uri)
+end
+for i = 1, 100 do
+    _M["print_uri_detailed_" .. i] = print_uri_detailed
+end
```
Comment (soulbird) [inline t/lib/server.lua:386]: Why do you need to export so many functions？

File: t/plugin/proxy-rewrite3.t [hunk 2]
```diff
@@ -200,3 +200,103 @@ passed
...
+
+
+=== TEST 11: unsafe uri not normalized at request
+--- request
+GET /print%5Furi%5Fdetailed%5F99 HTTP/1.1
+--- response_body
+ngx.var.uri: /print_uri_detailed_99
+ngx.var.request_uri: /print%5Furi%5Fdetailed%5F99
+--- no_error_log
+[error]
```
Comment (soulbird) [inline t/plugin/proxy-rewrite3.t:302]: Need test cases to cover the uri reaching upstream?

File: apisix/plugins/proxy-rewrite.lua [hunk 3]
```diff
@@ -161,7 +166,9 @@ function _M.rewrite(conf, ctx)
     end
 
     local upstream_uri = ctx.var.uri
-    if conf.uri ~= nil then
+    if conf.use_real_request_uri_unsafe then
+        upstream_uri = core.utils.resolve_var("$real_request_uri", ctx.var)
```
Comment (tzssangglass) [inline apisix/plugins/proxy-rewrite.lua:170]: we can use `core.utils.resolve_var("real_request_uri", ctx.var)` here? no `$`

File: t/plugin/proxy-rewrite3.t [hunk 4]
```diff
@@ -200,3 +200,103 @@ passed
...
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                 ngx.HTTP_PUT,
+                 [[{
+                        "methods": ["GET"],
+                        "plugins": {
+                            "proxy-rewrite": {
+                                "use_real_request_uri_unsafe": false
```
Comment (tzssangglass) [inline t/plugin/proxy-rewrite3.t:217]: It looks like we don't need this test case("use_real_request_uri_unsafe": false) because it defaults to false and there are existing test cases that override the default behavior.

File: t/plugin/proxy-rewrite3.t (hunk 2 above)
Comment (tokers) [reply t/plugin/proxy-rewrite3.t:302]: @soulbird The response body is returned by the upstream.

File: apisix/plugins/proxy-rewrite.lua [hunk 5]
```diff
@@ -161,7 +166,9 @@ function _M.rewrite(conf, ctx)
     end
 
     local upstream_uri = ctx.var.uri
-    if conf.uri ~= nil then
+    if conf.use_real_request_uri_unsafe then
+        upstream_uri = core.utils.resolve_var("real_request_uri", ctx.var)
```
Comment (spacewander) [inline apisix/plugins/proxy-rewrite.lua:170]: "real_request_uri" is a typo?

Comment (tzssangglass) [reply apisix/plugins/proxy-rewrite.lua:170]: He originally wrote `$real_request_uri`, but I suggested changing it to `real_request_uri` because `core.utils.resolve_var` doesn't seem to need `$` prefix?

Comment (ilteriseroglu-ty) [reply apisix/plugins/proxy-rewrite.lua]: Updated to use `ctx.var.real_request_uri`

File: docs/en/latest/plugins/proxy-rewrite.md [hunk 6]
```diff
@@ -33,14 +33,15 @@ The `proxy-rewrite` Plugin rewrites Upstream proxy information such as `scheme`,
...
+| use_real_request_uri_unsafe | boolean       | False    | false   |                                                                                                                                        | Use real_request_uri (originally request_uri in nginx) to bypass URI normalization. **Enabling this is considered unsafe as it bypasses all URI normalization steps**.                                                                                                                                                                                                                                                                                                     |
```
Comment (spacewander) [inline docs/en/latest/plugins/proxy-rewrite.md:44]:
```suggestion
| use_real_request_uri_unsafe | boolean       | False    | false   |                                                                                                                                        | Use real_request_uri (original $request_uri in nginx) to bypass URI normalization. **Enabling this is considered unsafe as it bypasses all URI normalization steps**.                                                                                                                                                                                                                                                                                                     |
```

File: t/plugin/proxy-rewrite3.t [hunk 7]
```diff
@@ -200,3 +200,103 @@ passed
...
+                }]]
+                )
+
+            if code >= 300 then
+                ngx.status = code
+            end
+            ngx.say(body)
+        }
+    }
+--- request
```
Comment (spacewander) [inline t/plugin/proxy-rewrite3.t:236]: We can remove `--- request` & `--- no_error_log` as they are already defined at the top
https://github.com/apache/apisix/blob/11bba0059c2a714486315c7beaf886eddd75ba60/t/plugin/proxy-rewrite3.t#L28-L34

**Review by spacewander — APPROVED** [review]

**Review by tzssangglass — APPROVED** [review]

**Review by soulbird — APPROVED** [review]
